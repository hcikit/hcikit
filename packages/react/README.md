# @hcikit/react

This package is a package that allows you to create and run HCI experiments. Visit hcikit.github.io for more information.
