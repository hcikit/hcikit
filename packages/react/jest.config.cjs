/** @type {import('@ts-jest/dist/types').InitialOptionsTsJest} */
module.exports = {
  preset: "ts-jest",
  testEnvironment: "jsdom",
  moduleNameMapper: {
    "@hcikit/workflow(.*)$": "<rootDir>/../workflow/src$1",
  },
  testMatch: ["./src/**/*.[jt]s?(x)", "**/?(*.)+(spec|test).[t]s?(x)"],
  modulePathIgnorePatterns: ["./dist/", "./node_modules/"],
  testPathIgnorePatterns: ["./dist/", "./node_modules/"],
  transformIgnorePatterns: ["./dist/", "/node_modules/"],
  setupFilesAfterEnv: ["<rootDir>/setupTests.ts"],
  collectCoverageFrom: [
    "src/**/*.{ts,tsx}",
    "!/node_modules/",
    "!src/setupTests.ts",
    "!src/**/*.stories.{js,ts,tsx,jsx}",
  ],
};
