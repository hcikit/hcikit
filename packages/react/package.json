{
  "name": "@hcikit/react",
  "version": "1.0.0-alpha.13",
  "description": "An experiment runner with a set of pre-built experiment components for running hci experiments",
  "author": "blainelewis1",
  "license": "MIT",
  "repository": "hcikit/hcikit",
  "module": "dist/index.js",
  "main": "dist/index.js",
  "type": "module",
  "keywords": [
    "react",
    "hci",
    "experiment"
  ],
  "types": "./dist/index.d.ts",
  "exports": {
    "import": "./dist/index.js",
    "default": "./dist/index.js"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "dist",
    "README.md"
  ],
  "engines": {
    "node": ">=8",
    "npm": ">=5"
  },
  "scripts": {
    "test": "NODE_OPTIONS=--experimental-vm-modules yarn jest --env=jsdom",
    "test:watch": "NODE_OPTIONS=--experimental-vm-modules yarn jest --watch --env=jsdom --colors --verbose",
    "start": "tsc --watch",
    "build": "tsc --build --verbose",
    "prepack": "yarn run build",
    "storybook": "start-storybook -p 6006",
    "build-storybook": "build-storybook"
  },
  "peerDependencies": {
    "@mui/icons-material": "^5.5.0",
    "@mui/material": "^5.5.0",
    "@mui/styles": "^5.5.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },
  "dependencies": {
    "@hcikit/workflow": "workspace:*",
    "core-js": "^3.26.1",
    "lodash-es": "^4.17.21",
    "prop-types": "^15.8.1",
    "react-error-boundary": "^3.1.4",
    "react-is": "^18.2.0",
    "react-markdown": "^8.0.3",
    "serialize-error": "^11.0.0",
    "styled-components": "^5.3.6"
  },
  "gitHead": "8d0f2af41121f70996e8b5803d006832b147fd86",
  "devDependencies": {
    "@testing-library/jest-dom": "^5.16.5",
    "@testing-library/react": "^13.4.0",
    "@types/jest": "^29.2.3",
    "@types/react": "^18.0.25",
    "@types/styled-components": "^5.1.26",
    "@types/testing-library__jest-dom": "^5.14.5",
    "jest-diff": "^29.3.1",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-scripts": "^5.0.1",
    "typescript": "^4.9.3"
  }
}
