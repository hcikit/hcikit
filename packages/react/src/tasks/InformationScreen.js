import React, { useEffect } from "react";
import Markdown from "markdown-to-jsx";

import { Button } from "@material-ui/core";
import { CenteredNicePaper } from "../components";
import PropTypes from "prop-types";

/**
 * Creates a screen with information or instructions and a continue button.
 * Accepts Markdown.
 * By setting withContinue to false it can also be used as an end screen of an experiment.
 * Uses a keyboard shortcut when specified.
 */
const InformationScreen = ({
  content,
  withContinue = true,
  taskComplete,
  centerX,
  centerY,
  shortcutEnabled,
  key = "Enter",
}) => {
  useEffect(() => {
    function handleKeyDown(e) {
      if (e.key === key && shortcutEnabled && withContinue) {
        taskComplete();
      }
    }
    window.addEventListener("keydown", handleKeyDown);
    return () => window.removeEventListener("keydown", handleKeyDown);
  });

  return (
    <CenteredNicePaper centerX={centerX} centerY={centerY}>
      <Markdown>{content}</Markdown>
      <br />
      {withContinue && (
        <Button color="primary" onClick={taskComplete}>
          Continue {shortcutEnabled && `(${key})`}
        </Button>
      )}
    </CenteredNicePaper>
  );
};

InformationScreen.propTypes = {
  centerX: PropTypes.bool,
  centerY: PropTypes.bool,
  content: PropTypes.string.isRequired,
  taskComplete: PropTypes.func,
  withContinue: PropTypes.bool,
  shortcutEnabled: PropTypes.bool,
  /* The KeyboardEvent.key value to use as a shortcut */
  key: PropTypes.string,
};

export default InformationScreen;
