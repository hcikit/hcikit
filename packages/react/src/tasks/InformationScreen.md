```js
<InformationScreen
  content={`
# Hello World

Information screens can be written with markdown.
`}
/>
```
