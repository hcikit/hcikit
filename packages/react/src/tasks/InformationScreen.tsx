import React, { useEffect } from "react";
import Markdown from "markdown-to-jsx";

import { Button } from "@material-ui/core";
import { CenteredNicePaper } from "../components";
import PropTypes from "prop-types";
import { useExperiment } from "../core/Experiment";

/**
 * Creates a screen with information or instructions and a continue button.
 * Accepts Markdown.
 * By setting withContinue to false it can also be used as an end screen of an experiment.
 * Uses a keyboard shortcut when specified.
 */
const InformationScreen: React.FunctionComponent<{
  content: string;
  withContinue?: boolean;
  centerX?: boolean;
  centerY?: boolean;
  shortcutEnabled?: boolean;
  shortcut?: string;
}> = ({
  content,
  withContinue = true,
  centerX,
  centerY,
  shortcutEnabled,
  shortcut = "Enter",
}) => {
  const experiment = useExperiment();

  useEffect(() => {
    function handleKeyDown(e: KeyboardEvent) {
      if (e.key === shortcut && shortcutEnabled && withContinue) {
        experiment.taskComplete();
      }
    }
    window.addEventListener("keydown", handleKeyDown);
    return () => window.removeEventListener("keydown", handleKeyDown);
  }, [experiment, shortcut, shortcutEnabled, withContinue]);

  return (
    <CenteredNicePaper centerX={centerX} centerY={centerY}>
      <Markdown>{content}</Markdown>
      <br />
      {withContinue && (
        <Button color="primary" onClick={experiment.taskComplete}>
          Continue {shortcutEnabled && `(${shortcut})`}
        </Button>
      )}
    </CenteredNicePaper>
  );
};

InformationScreen.propTypes = {
  centerX: PropTypes.bool,
  centerY: PropTypes.bool,
  content: PropTypes.string.isRequired,
  withContinue: PropTypes.bool,
  shortcutEnabled: PropTypes.bool,
  /* The KeyboardEvent.key value to use as a shortcut */
  shortcut: PropTypes.string,
};

export default InformationScreen;
