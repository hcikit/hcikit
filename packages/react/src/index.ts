import { PropsWithChildren, ReactElement, Component } from "react";
import { ControlFunctions } from "./core/Experiment.js";
import * as allTasks from "./tasks/index.js";

export * from "./tasks/index.js";
export * from "./components/index.js";

export { allTasks };

export { withGridItem } from "./GridLayout.js";

export { default as GridLayout } from "./GridLayout.js";
export { default } from "./core/Experiment.js";
export * from "./core/Experiment.js";
export {
  BasePersistence,
  StoragePersistence,
  FileSystemPersistence,
} from "./persistence/index.js";

type TaskProps = ControlFunctions;
type StaticProps = {
  layoutProps?: Record<string, any>;
};

export type Task<P = {}> = React.ComponentType<TaskProps & P> & StaticProps;

export interface FunctionTask<P = {}>
  extends React.FunctionComponent<P & ControlFunctions> {
  (
    props: PropsWithChildren<P & ControlFunctions>,
    context?: any
  ): ReactElement<any, any> | null;
}

export abstract class ClassTask<P = {}, S = {}, SS = any> extends Component<
  P & TaskProps,
  S,
  SS
> {}
