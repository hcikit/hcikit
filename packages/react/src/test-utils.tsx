import { render, RenderResult } from "@testing-library/react";

import {
  ConfigContext,
  ConfigMutatorContext,
  ControlFunctions,
} from "./core/Experiment.js";
import { Configuration, ExperimentIndex } from "@hcikit/workflow";
import React from "react";
import { BasePersistence } from "./persistence/index.js";

class DoNothingPersistence extends BasePersistence {
  load(): Promise<Configuration | undefined> {
    return Promise.resolve(undefined);
  }
  save(state: Configuration): Promise<void> {
    return Promise.resolve();
  }
  clear(): Promise<void> {
    return Promise.resolve();
  }
  init(): Promise<void> {
    return Promise.resolve();
  }
}

const identities: ControlFunctions = {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  advance: (_index?: ExperimentIndex) => {
    // do nothing.
  },
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  log: (_log: unknown) => {
    // do nothing.
  },
  modify: (
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    _modifiedConfig: Record<string, unknown>,
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    _index?: ExperimentIndex
  ) => {
    // do nothing.
  },
  persistence: new DoNothingPersistence(),
};

export const renderWithProvider = (
  Component: React.ReactElement,
  funcs: Partial<ControlFunctions> = identities,
  configuration: Configuration = {}
): RenderResult => {
  return render(
    <ConfigContext.Provider value={configuration}>
      <ConfigMutatorContext.Provider value={{ ...identities, ...funcs }}>
        {Component}
      </ConfigMutatorContext.Provider>
    </ConfigContext.Provider>
  );
};
