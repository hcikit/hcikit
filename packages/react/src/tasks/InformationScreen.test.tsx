import React from "react";
import InformationScreen from "./InformationScreen.js";
import { fireEvent, screen } from "@testing-library/react";

import { renderWithProvider } from "../test-utils.js";
import userEvent from "@testing-library/user-event";
import { jest } from "@jest/globals";

describe("InformationScreen", () => {
  it("renders without crashing", () => {
    renderWithProvider(<InformationScreen content="hello" />);
  });

  it("renders without continue", () => {
    renderWithProvider(
      <InformationScreen withContinue={false} content="hello" />
    );
    expect(screen.queryByText("Continue")).not.toBeInTheDocument();
  });

  it("advances to the next task", async () => {
    const advance = jest.fn();
    renderWithProvider(<InformationScreen content="Hello World" />, {
      advance,
    });

    await userEvent.click(screen.getByText("Continue"));
    expect(advance).toBeCalledTimes(1);
  });

  it("doesn't advance when withContinue is false", () => {
    const advance = jest.fn();

    const rendered = renderWithProvider(
      <InformationScreen
        withContinue={false}
        shortcutEnabled
        content="Hello World"
      />,
      { advance }
    );

    // TODO: this test is likely broken atm, missing an act
    fireEvent.keyDown(rendered.baseElement, { key: "Enter", code: "Enter" });

    expect(advance).not.toHaveBeenCalled();
  });

  it("advances with a keyboard press", () => {
    const advance = jest.fn();
    const rendered = renderWithProvider(
      <InformationScreen shortcutEnabled content="Hello World" />,
      { advance: advance }
    );

    fireEvent.keyDown(rendered.baseElement, { key: "Enter", code: "Enter" });
    expect(advance).toBeCalledTimes(1);
  });

  it("renders markdown properly", () => {
    const markdown = `
# Hello World

 - This is some markdown
 - that should render *properly*
`;

    expect(
      renderWithProvider(<InformationScreen content={markdown} />).asFragment()
    ).toMatchSnapshot();
  });
});
