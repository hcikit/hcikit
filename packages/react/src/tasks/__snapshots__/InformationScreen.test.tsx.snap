// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InformationScreen renders markdown properly 1`] = `
<DocumentFragment>
  <div
    class="sc-gKPRtg cmBYij"
  >
    <div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 sc-jSUZER goswre css-1ps6pg7-MuiPaper-root"
      >
        <h1>
          Hello World
        </h1>
        

        <ul>
          

          <li>
            This is some markdown
          </li>
          

          <li>
            that should render 
            <em>
              properly
            </em>
          </li>
          

        </ul>
        <br />
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Continue 
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
